Data-Structure-Linked-Lists.md

```diff
@@ -0,0 +1,465 @@
+
+    For this application, there should be no NULL pointers unless there is absolutely no one requesting CPU time, i.e list is empty.
+
+## Basic Operations
+
+1. Insertion
+
+    To add a new element to the list.
+
+    ```
+
+    Insertion at the beginning
+    ------------------------
+
+    * Create a new node with given data.
+    * Point new node's `next` to old `head`.
+    * Point `head` to this new node.
+
+    Insertion in the middle/end
+    --------------------------
+    Insertion after node X.
+
+    * Create a new node with given data.
+    * Point new node's `next` to old X's `next`.
+    * Point X's `next` to this new node.
+
+    ```
+    Time Complexity: O(1)
+
+2. Deletion
+
+    To delete existing element from the list.
+
+    ```
+
+    Deletion at the beginning
+    -----------------------
+
+    * Get the node pointed by `head` as Temp.
+    * Point `head` to Temp's `next`.
+    * Free memory used by Temp node.
+
+    Deletion in the middle/end
+    -------------------------
+    Deletion after node X.
+
+    * Get the node pointed by `X` as Temp.
+    * Point X's `next` to Temp's `next`.
+    * Free memory used by Temp node.
+
+    ```
+    Time Complexity: O(1)
+
+3. Traversing
+
+    To travel acroos the list.
+
+    ```
+
+    Traversal
+    --------
+
+    * Get the node pointed by `head` as Current.
+    * Check if Current is not null and display it.
+    * Point Current to Current's `next` and move to above step.
+
+    ```
+    Time Complexity: O(n) // Here n is size of link-list
+
+## Implementation
+
+### C++ implementation of singly linked list
+
+```cpp
+
+// Header files
+#include <iostream>
+
+struct node
+{
+    int data;
+    struct node *next;
+};
+
+// Head pointer always points to first element of the linked list
+struct node *head = NULL;
+
+```
+
+#### Printing data in each node
+
+```cpp
+
+// Display the list
+void printList()
+{
+    struct node *ptr = head;
+
+    // Start from the beginning
+    while(ptr != NULL)
+    {
+        std::cout << ptr->data << " ";
+        ptr = ptr->next;
+    }
+
+    std::cout << "\n";
+}
+
+```
+
+#### Insertion at the beginning
+
+```cpp
+
+// Insert link at the beginning
+void insertFirst(int data)
+{
+    // Create a new node
+    struct node *new_node = new struct node;
+
+    new_node->data = data;
+
+    // Point it to old head
+    new_node->next = head;
+
+    // Point head to new node
+    head = new_node;
+
+    std::cout << "Inserted successfully\n";
+}
+
+```
+
+#### Deletion at the beginning
+
+```cpp
+
+// Delete first item
+void deleteFirst()
+{
+    // Save reference to head
+    struct node *temp = head;
+
+    // Point head to head's next
+    head = head->next;
+
+    // Free memory used by temp
+    delete temp;
+
```

Comment:
`temp = null` its a standard practice to follow. 
The memory may be de-allocated but the location can still be referenced, hence setting to `null` is a safe practice